go-controller/pkg/kube/annotator_test.go

```diff
@@ -0,0 +1,605 @@
+	nodeAnnotator := nodeAnnotator{}
+	nodeAnnotator.changes = make(map[string]*action)
+	tests := []struct {
+		desc string
+		key  string
+	}{
+		{
+			desc: "Positive test code path",
+			key:  "key",
+		},
+	}
+	for i, tc := range tests {
+		t.Run(fmt.Sprintf("%d:%s", i, tc.desc), func(t *testing.T) {
+
+			nodeAnnotator.Delete(tc.key)
+
+		})
+	}
+}
+
+func TestNodeAnnotator_Run(t *testing.T) {
+	mockIface := new(mock_clientgo_iface.Interface)
+	mockCoreIface := new(mock_core_v1.CoreV1Interface)
+	mockNodeIface := new(mock_core_v1.NodeInterface)
+	nodeAnnotator := nodeAnnotator{}
+	nodeAnnotator.kube = &Kube{KClient: mockIface}
+	nodeAnnotator.changes = make(map[string]*action)
+	nodeAnnotator.node = &v1.Node{}
+	nodeAnnotator.node.Annotations = make(map[string]string)
+	nodeAnnotator.node.Annotations["key"] = ""
+	tests := []struct {
+		desc                   string
+		expectedErr            error
+		onRetMockInterfaceArgs *ovntest.TestifyMockHelper
+		onRetMockCoreArgs      *ovntest.TestifyMockHelper
+		onRetMockNodeArgs      *ovntest.TestifyMockHelper
+		act                    *action
+	}{
+		{
+			desc:        "Test code path when length of annotations is zero",
+			expectedErr: nil,
+		},
+		{
+			desc:                   "Positive test code path",
+			expectedErr:            nil,
+			onRetMockInterfaceArgs: &ovntest.TestifyMockHelper{"CoreV1", []string{}, []interface{}{mockCoreIface}},
+			onRetMockCoreArgs:      &ovntest.TestifyMockHelper{"Nodes", []string{"string"}, []interface{}{mockNodeIface}},
+			onRetMockNodeArgs:      &ovntest.TestifyMockHelper{"Patch", []string{"*context.emptyCtx", "string", "types.PatchType", "[]uint8", "v1.PatchOptions"}, []interface{}{nil, nil}},
+			act:                    &action{key: "key", origVal: "val"},
+		},
+		{
+			desc:                   "Test code path when origVal is nil",
+			expectedErr:            nil,
+			onRetMockInterfaceArgs: &ovntest.TestifyMockHelper{"CoreV1", []string{}, []interface{}{mockCoreIface}},
+			onRetMockCoreArgs:      &ovntest.TestifyMockHelper{"Nodes", []string{"string"}, []interface{}{mockNodeIface}},
+			onRetMockNodeArgs:      &ovntest.TestifyMockHelper{"Patch", []string{"*context.emptyCtx", "string", "types.PatchType", "[]uint8", "v1.PatchOptions"}, []interface{}{nil, nil}},
+			act:                    &action{key: "key"},
+		},
+		{
+			desc:                   "Negative test code path",
+			expectedErr:            fmt.Errorf("mock Error: Failed to run SetAnnotationsOnNode"),
+			onRetMockInterfaceArgs: &ovntest.TestifyMockHelper{"CoreV1", []string{}, []interface{}{mockCoreIface}},
+			onRetMockCoreArgs:      &ovntest.TestifyMockHelper{"Nodes", []string{"string"}, []interface{}{mockNodeIface}},
+			onRetMockNodeArgs:      &ovntest.TestifyMockHelper{"Patch", []string{"*context.emptyCtx", "string", "types.PatchType", "[]uint8", "v1.PatchOptions"}, []interface{}{nil, fmt.Errorf("mock Error: Failed to run SetAnnotationsOnNode")}},
```

Comment:
it would be good to have an `action` object with the `failFn` set to be non nil

Reply:
Thanks for the feedback.
What would a non nil value for `failFn` look like?

Reply:
you will have a support function defined as below in the annotator_test.go file

```
func supportfailfnImpl(obj interface{}, key string, val interface{}) {
	// Do nothing as it does not return anything
}
```

and have one of the test table struct values populated as below

```
act:                    &action{key: "key", failFn: supportfailfnImpl},
```

Reply:
Refer the link https://stackoverflow.com/questions/9398739/working-with-function-types-in-go for better illustration